File: KantaroBLL/Canzone/Canzone.vb
Public Class Canzone

    Public Property VariazioneInSemitoni As Integer
    Public Property Tag As Object

    Public Shared Async Function FromStreamAsync(stream As System.IO.StreamReader) As Task(Of Canzone)
        If stream Is Nothing Then Throw New ArgumentNullException("stream")
        Return FromXDocument(Await LoadXDocumentAsync(stream).ConfigureAwait(False))
    End Function

    Private Shared Async Function LoadXDocumentAsync(stream As System.IO.StreamReader) As Task(Of XDocument)
        If stream Is Nothing Then Throw New ArgumentNullException("stream")
        Dim t As New Task(Of XDocument)(Function()
                                            Try
                                                Return XDocument.Load(stream)
                                            Catch ex As Exception
                                                Return Nothing
                                            End Try
                                        End Function)
        t.Start()
        Return Await t.ConfigureAwait(False)
    End Function

    Public Shared Function FromXDocument(document As XDocument) As Canzone
        If document Is Nothing Then Return Nothing
        Try
            Dim xSong = document
            Dim song As New Canzone
            With song
                .Titolo = xSong.<canzone>.@title
                .Autore = xSong.<canzone>.@autore
                .VariazioneInSemitoni = xSong.<canzone>.@variazione
                For Each st In xSong.<canzone>.<strofa>
                    If Not String.IsNullOrWhiteSpace(st.@ref) Then
                        .Strofe.Add(New StrofaRipetuta With {.Riferimento = st.@ref})
                        'If .NamedParagraphs.ContainsKey(st.@ref) Then
                        '    .Strofe.Add(.NamedParagraphs(st.@ref).Clone())
                        'End If
                        Continue For
                    End If
                    Dim strofa As New Strofa
                    For Each elem In st.Elements
                        Dim p As Parte = Nothing
                        If elem.Name = "parte" Then
                            p = New Parte()
                            If Not String.IsNullOrEmpty(elem.@accordo) Then p.Accordo = CType(elem.@accordo, Accordo)
                            If Not elem.IsEmpty Then p.Testo = elem.Value
                        End If
                        strofa.Parti.Add(p)
                    Next
                    If Not String.IsNullOrWhiteSpace(st.@name) Then
                        strofa.Nome = st.@name
                        '.NamedParagraphs.Add(st.@name, strofa)
                    End If
                    .Strofe.Add(strofa)
                Next
            End With
            Return song
        Catch ex As Exception
            Throw New Exception(ex.Message, ex) With {.Source = document.ToString}
        End Try
    End Function

    Public Function ToXml() As XDocument
        Dim d As New XDocument()
        Dim root = <canzone/>
        If Me.Titolo IsNot Nothing Then root.Add(New XAttribute("title", Me.Titolo))
        If Me.Autore IsNot Nothing Then root.Add(New XAttribute("autore", Me.Autore))
        If Me.VariazioneInSemitoni <> 0 Then root.Add(New XAttribute("variazione", Me.VariazioneInSemitoni))
        For Each st In Strofe
            Dim strofa = <strofa/>
            If TypeOf st Is StrofaRipetuta Then
                strofa.Add(New XAttribute("ref", DirectCast(st, StrofaRipetuta).Riferimento))
            Else
                For Each p In st.Parti
                    If p Is Nothing Then
                        strofa.Add(<br/>)
                    Else
                        Dim parte = <parte/>
                        If p.Accordo IsNot Nothing Then
                            parte.Add(New XAttribute("accordo", p.Accordo.ToString))
                        End If
                        If p.Testo IsNot Nothing Then
                            parte.Value = p.Testo
                        End If
                        strofa.Add(parte)
                    End If
                Next
            End If
            root.Add(strofa)
        Next
        d.Add(root)
        Return d
    End Function

    Public Property Titolo As String
    Public Property Autore As String

    ''' <summary>
    ''' Descrizione della canzone (costituita dal titolo + l'autore se presente)
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property Description As String
        Get
            Dim sb As New Text.StringBuilder
            If Titolo IsNot Nothing Then sb.Append(Titolo)
            If Not String.IsNullOrWhiteSpace(Autore) Then
                sb.Append(" (")
                sb.Append(Autore)
                sb.Append(")")
            End If
            Return sb.ToString
        End Get
    End Property

    Private _strofe As IList(Of Strofa)
    Public ReadOnly Property Strofe As IList(Of Strofa)
        Get
            If _strofe Is Nothing Then
                _strofe = New List(Of Strofa)
            End If
            Return _strofe
        End Get
    End Property

    Public Function GetStrofe() As IList(Of Strofa)
        Dim ret As New List(Of Strofa)
        For Each s In Strofe
            Dim n As Strofa = DirectCast(s, ICloneable).Clone
            If VariazioneInSemitoni <> 0 Then
                Dim dist As New Distanza With {.Semitoni = VariazioneInSemitoni}
                For Each p In n.Parti
                    If p Is Nothing Then Continue For
                    If p.Accordo IsNot Nothing Then
                        p.Accordo = p.Accordo + dist
                    End If
                Next
            End If
            ret.Add(n)
        Next
        Return ret
    End Function

    Public Overloads Shared Function Equals(obj1 As Object, obj2 As Object) As Boolean
        If obj1 Is Nothing AndAlso obj2 Is Nothing Then Return True
        If Not TypeOf obj1 Is Canzone Then Return False
        Return DirectCast(obj1, Canzone).Equals(obj2)
    End Function

    Public Overrides Function Equals(obj As Object) As Boolean
        If obj Is Nothing Then Return False
        If Not TypeOf obj Is Canzone Then Return False
        Dim b = DirectCast(obj, Canzone)
        Dim uguali = String.Equals(Me.Autore, b.Autore) AndAlso
                     String.Equals(Me.Titolo, b.Titolo) AndAlso
                     Integer.Equals(Me.VariazioneInSemitoni, b.VariazioneInSemitoni)
        If uguali Then uguali = (Me.Strofe.Count = b.Strofe.Count)
        If uguali Then
            For i = 0 To Strofe.Count - 1
                uguali = Strofe(i).Equals(b.Strofe(i))
                If Not uguali Then Exit For
            Next
        End If
        Return uguali
    End Function

    Public Shared Operator <>(ByVal left As Canzone, ByVal right As Canzone) As Boolean
        Return Not Canzone.Equals(left, right)
    End Operator

    Public Shared Operator =(ByVal left As Canzone, ByVal right As Canzone) As Boolean
        Return Canzone.Equals(left, right)
    End Operator
End Class

File: Enzo.Music/Canzone/Canzone.cs
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Numerics;
using System.Reflection.Metadata.Ecma335;
using System.Runtime.CompilerServices;
using System.Text;
using System.Xml.Linq;
using static System.Net.Mime.MediaTypeNames;

namespace Enzo.Music;
public class Canzone : IEquatable<Canzone>, IEqualityOperators<Canzone, Canzone, bool>
{
    public int VariazioneInSemitoni { get; set; }
    public object? Tag { get; set; }
    public string? Titolo { get; set; }
    public string? Autore { get; set; }

    public readonly List<Strofa> Strofe = [];

    /// <summary>
    /// Descrizione della canzone (costituita dal titolo + l'autore se presente)
    /// </summary>
    public string Description
    {
        get
        {
            StringBuilder sb = new();
            if (Titolo != null) sb.Append(Titolo);
            if (!string.IsNullOrWhiteSpace(Autore))
                sb.Append($" ({Autore})");
            return sb.ToString();
        }
    }

    public static async Task<Canzone?> FromStreamAsync(StreamReader stream)
    {
        ArgumentNullException.ThrowIfNull(stream);
        return FromXDocument(await LoadXDocumentAsync(stream).ConfigureAwait(false));
    }

    public static async Task<XDocument?> LoadXDocumentAsync(StreamReader stream)
    {
        ArgumentNullException.ThrowIfNull(stream);
        var t = new Task<XDocument?>(() =>
        {
            try
            {
                return XDocument.Load(stream);
            }
            catch// (Exception ex)
            {
                return null;
            }
        });
        t.Start();
        return await t.ConfigureAwait(false);
    }

    public static Canzone? FromXDocument(XDocument? document)
    {
        if (document is null) return null;
        try
        {
            Canzone song = new()
            {
                Titolo = document.Element("Canzone")?.Attribute("Titolo")?.Value ?? string.Empty,
                Autore = document.Element("Canzone")?.Attribute("Autore")?.Value ?? string.Empty
            };
            if (int.TryParse(document.Element("Canzone")?.Attribute("variazione")?.Value, out int variazione))
                song.VariazioneInSemitoni = variazione;
            foreach (XElement st in document.Element("canzone")?.Elements("strofa") ?? [])
            {
                if (!string.IsNullOrWhiteSpace(st.Attribute("ref")?.Value))
                {
                    song.Strofe.Add(new StrofaRipetuta(st.Attribute("ref")?.Value ?? string.Empty));
                    continue;
                }
                Strofa strofa = new();
                foreach (XElement elem in st.Elements())
                {
                    Parte? p = null;
                    if (elem.Name == "parte")
                    {
                        p = new Parte();
                        if (!string.IsNullOrWhiteSpace(elem.Attribute("accordo")?.Value))
                            p.Accordo = (Accordo?)(elem.Attribute("accordo")?.Value ?? string.Empty);
                        if (!elem.IsEmpty) p.Testo = elem.Value;
                    }
                    strofa.Parti.Add(p);
                }
                if (!string.IsNullOrWhiteSpace(st.Attribute("name")?.Value))
                    strofa.Nome = st.Attribute("name")?.Value;
                song.Strofe.Add(strofa);
            }
            return song;
        }
        catch (Exception ex)
        {
            throw new Exception(ex.Message, ex) { Source = document.ToString() };
        }
    }

    public XDocument ToXml()
    {
        XDocument d = new();
        XElement root = new("canzone");
        if (Titolo != null) root.Add(new XAttribute("title", Titolo));
        if (Autore != null) root.Add(new XAttribute("autore", Autore));
        if (VariazioneInSemitoni != 0) root.Add(new XAttribute("variazione", VariazioneInSemitoni));
        foreach (var st in Strofe)
        {
            XElement strofa = new("strofa");
            if (st is StrofaRipetuta stRip)
                strofa.Add(new XAttribute("ref", stRip.Riferimento));
            else
            {
                foreach (var p in st.Parti)
                {
                    if (p is null)
                        strofa.Add(new XElement("br"));
                    else
                    {
                        XElement parte = new("parte");
                        if (p.Accordo != null)
                            parte.Add(new XAttribute("accordo", p.Accordo.ToString()));
                        if (p.Testo != null)
                            parte.Value = p.Testo;
                        strofa.Add(parte);
                    }
                }
            }
            root.Add(strofa);
        }
        d.Add(root);
        return d;
    }

    public IList<Strofa> GetStrofe()
    {
        List<Strofa> ret = [];
        foreach (var s in Strofe)
        {
            Strofa n = (Strofa)s.Clone();
            if (VariazioneInSemitoni != 0)
            {
                Distanza dist = new() { Semitoni = VariazioneInSemitoni };
                foreach (var p in n.Parti)
                {
                    if (p is null) continue;
                    if (p.Accordo != null)
                        p.Accordo += dist;
                }
            }
            ret.Add(n);
        }
        return ret;
    }

    public bool Equals(Canzone? other)
    {
        if (other is null) return false;
        var uguali = string.Equals(Autore, other.Autore) &&
                     string.Equals(Titolo, other.Titolo) &&
                     VariazioneInSemitoni == other.VariazioneInSemitoni;
        if (uguali) uguali = Strofe.Count == other.Strofe.Count;
        if (uguali)
        {
            for (var i = 0; i < Strofe.Count; i++)
            {
                uguali = Strofe[i].Equals(other.Strofe[i]);
                if (!uguali) break;
            }
        }
        return uguali;
    }

    public static bool Equals(Canzone? obj1, Canzone? obj2)
    {
        if (obj1 is null || obj2 is null) return false;
        return obj1.Equals(obj2);
    }

    public override bool Equals(object? obj)
    {
        if (obj is not Canzone) return false;
        return Equals((Canzone?)obj);
    }

    public static bool operator ==(Canzone? left, Canzone? right)
    {
        if (left is null || right is null) return false;
        return left.Equals(right);
    }

    public static bool operator !=(Canzone? left, Canzone? right)
    {
        if (left is null || right is null) return false;
        return !left.Equals(right);
    }

    public override int GetHashCode()
    {
        var hc = (Titolo??string.Empty).GetHashCode() 
            ^ (Autore??string.Empty).GetHashCode() 
            ^ VariazioneInSemitoni.GetHashCode();
        hc ^= Strofe.GetHashCode();
        return hc;
    }
}
